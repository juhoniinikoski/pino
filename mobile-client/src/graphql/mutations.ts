import { gql } from '@apollo/client';

export const AUTHORIZE = gql`
  mutation authorize($email: String!, $password: String!) {
    authorize(credentials: { email: $email, password: $password }) {
      accessToken
    }
  }
`;

export const FOLLOW_COLLECTION = gql`
  mutation followCollection($collectionId: ID!) {
    followCollection(collectionId: $collectionId)
  }
`;
