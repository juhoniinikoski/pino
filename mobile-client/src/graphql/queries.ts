import { gql } from '@apollo/client';

export const GET_CHANNELS = gql`
  query Channels($first: Int, $after: String, $searchKeyword: String) {
    channels(first: $first, after: $after, searchKeyword: $searchKeyword) {
      edges {
        node {
          id
          name
          followedBy
          questions
        }
        cursor
      }
      totalCount
      pageInfo {
        endCursor
        startCursor
        hasNextPage
      }
    }
  }
`;

export const GET_USERS_CHANNELS = gql`
  query Channels($first: Int, $after: String, $searchKeyword: String) {
    channels(
      first: $first
      after: $after
      searchKeyword: $searchKeyword
      followedByAuthorized: true
    ) {
      edges {
        node {
          id
          name
          followedBy
          questions
        }
        cursor
      }
      totalCount
      pageInfo {
        endCursor
        startCursor
        hasNextPage
      }
    }
  }
`;

export const GET_QUESTIONS = gql`
  query Questions(
    $channelId: ID
    $stackId: ID
    $createdBy: ID
    $first: Int
    $after: String
  ) {
    questions(
      channelId: $channelId
      stackId: $stackId
      createdBy: $createdBy
      first: $first
      after: $after
    ) {
      edges {
        node {
          id
          question
          createdById
          type
          answers {
            answer
            correct
            id
          }
          updatedAt
          createdAt
        }
        cursor
      }
      totalCount
      pageInfo {
        startCursor
        endCursor
        hasNextPage
        hasPreviousPage
      }
    }
  }
`;

export const GET_STACKS = gql`
  query Stacks(
    $public: Boolean
    $createdBy: ID
    $first: Int
    $after: String
    $searchKeyword: String
  ) {
    stacks(
      first: $first
      after: $after
      createdBy: $createdBy
      searchKeyword: $searchKeyword
      public: $public
    ) {
      totalCount
      edges {
        node {
          id
          name
          public
          questions
          followedBy
          createdById
          createdAt
          updatedAt
          tags {
            id
            name
          }
        }
        cursor
      }
    }
  }
`;

export const GET_AUTHORIZED_USER = gql`
  query authorizedUser {
    authorizedUser {
      id
      email
    }
  }
`;
