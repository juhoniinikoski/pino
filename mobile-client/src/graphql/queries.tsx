import { gql } from '@apollo/client';

export const GET_CHANNELS = gql`
  query Channels($first: Int, $after: String, $searchKeyword: String) {
    channels(first: $first, after: $after, searchKeyword: $searchKeyword) {
      edges {
        node {
          id
          name
          followedBy
          questions
        }
        cursor
      }
      totalCount
      pageInfo {
        endCursor
        startCursor
        hasNextPage
      }
    }
  }
`;

export const GET_QUESTIONS = gql`
  query Questions(
    $channelId: ID
    $stackId: ID
    $createdBy: ID
    $first: Int
    $after: String
  ) {
    questions(
      channelId: $channelId
      stackId: $stackId
      createdBy: $createdBy
      first: $first
      after: $after
    ) {
      edges {
        node {
          id
          question
          createdById
          type
          answers {
            answer
            correct
            id
          }
          updatedAt
          createdAt
        }
        cursor
      }
      totalCount
      pageInfo {
        startCursor
        endCursor
        hasNextPage
        hasPreviousPage
      }
    }
  }
`;
