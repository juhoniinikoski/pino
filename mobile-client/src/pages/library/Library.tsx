import { FlatList, StyleSheet, Text } from 'react-native';
import * as React from 'react';
import Layout from '../../components/layout/Layout';
import parseNodes from '../../utils/parseNodes';
import { Channel, Stack } from '../../utils/types';
import ChannelBox from '../../components/channelBox/ChannelBox';
import useUserChannels from '../../hooks/useUserChannels';
import useStacks from '../../hooks/useStacks';
import StackBox from '../../components/stackBox/StackBox';

/* eslint-disable no-underscore-dangle */

const styles = StyleSheet.create({
  channelLoadText: {
    textAlign: 'center',
    textAlignVertical: 'center',
  },
  header: {
    marginTop: 16,
    marginBottom: 8,
  },
});

type DataType = Stack | Channel;

function isChannel(item: DataType): item is Channel {
  return item.__typename === 'Channel';
}

const Library = () => {
  const [followedChannels, setFollowedChannels] = React.useState<Channel[]>([]);
  const [userStacks, setUserStacks] = React.useState<Stack[]>([]);

  const [data, setData] = React.useState<DataType[]>([]);

  const { channels: followedRaw, loading: channelLoad } = useUserChannels();
  const { stacks: userStacksRaw, loading: stackLoad } = useStacks();

  React.useEffect(() => {
    const followed = followedRaw ? parseNodes<Channel>(followedRaw) : [];
    const userStacks = userStacksRaw ? parseNodes<Stack>(userStacksRaw) : [];
    setFollowedChannels(followed);
    setUserStacks(userStacks);
  }, [followedRaw, userStacksRaw, setFollowedChannels, setUserStacks]);

  React.useEffect(() => {
    if (!(channelLoad || stackLoad)) {
      setData([...followedChannels, ...userStacks]);
    }
  }, [followedChannels, userStacks, channelLoad, stackLoad]);

  if (
    (channelLoad || stackLoad) &&
    (!followedChannels.length || !userStacks.length)
  ) {
    return (
      <Layout>
        <Text style={styles.channelLoadText}>Loading</Text>
      </Layout>
    );
  }

  return (
    <Layout>
      <FlatList
      testID='library-list'
        data={data}
        keyExtractor={(item: DataType) => item.id + item.__typename}
        renderItem={({ item }) =>
          isChannel(item) ? (
            <ChannelBox channel={item} followedByUser />
          ) : (
            <StackBox stack={item} />
          )
        }
      />
    </Layout>
  );
};

export default Library;
