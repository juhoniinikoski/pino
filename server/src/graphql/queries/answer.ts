import { gql } from 'apollo-server';
import { getAnswer } from '../../services/answer/answerService';

export const typeDefs = gql`
  extend type Query {
    """
    Returns an answer.
    """
    answer(id: ID!): Answer!
  }
`;

interface Args {
  id: string | number;
}

export const resolvers = {
  Query: {
    answer: (_obj: null, { id }: Args) => getAnswer(id),
  },
};

export default {
  typeDefs,
  resolvers,
};
