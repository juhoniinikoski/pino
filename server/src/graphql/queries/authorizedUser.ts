import { gql } from 'apollo-server'
import AuthService from '../../services/authentication/authService'
import { getAuthorizedUser } from '../../services/user/userService'
import { Context } from '../../utils/entities'

export const typeDefs = gql`
  extend type Query {
    """
    Returns the authorized user.
    """
    authorizedUser: User
  }
`

export const resolvers = {
  Query: {
    authorizedUser: (_obj: null, _args: null, { authService }: Context) => getAuthorizedUser(authService)
  },
}

export default {
  typeDefs,
  resolvers,
}