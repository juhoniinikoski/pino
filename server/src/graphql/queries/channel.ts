import { gql } from 'apollo-server';
import { getChannel } from '../../services/channel/channelService';

export const typeDefs = gql`
  extend type Query {
    """
    Returns a stack and related questions.
    """
    channel(id: ID!): Channel!
  }
`;

interface Args {
  id: string | number;
}

export const resolvers = {
  Query: {
    stack: async (_obj: null, { id }: Args) => getChannel(id),
  },
};

export default {
  typeDefs,
  resolvers,
};
