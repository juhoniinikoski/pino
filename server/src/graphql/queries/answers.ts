import { gql } from 'apollo-server';
import { getAnswers } from '../../services/answer/answerService';

export const typeDefs = gql`
  extend type Query {
    """
    Returns list of answers.
    """
    answers: [Answer!]
  }
`;

export const resolvers = {
  Query: {
    answers: async () => getAnswers(),
  },
};

export default {
  typeDefs,
  resolvers,
};
