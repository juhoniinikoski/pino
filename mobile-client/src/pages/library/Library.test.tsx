import * as React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { render, RenderAPI, waitFor } from '@testing-library/react-native';
import { GET_CHANNELS, GET_STACKS } from '../../graphql/queries';
import Library from './Library';

jest.useFakeTimers()

/* eslint-disable no-unused-expressions */

const mocks = [
  {
    request: {
      query: GET_CHANNELS,
    },
    result: {
      data: [
        {
          cursor:
            'WyIyMDIyLTA0LTI4VDA4OjEyOjQ2LjI0MVoiLCJrYXVwcGlrc2VlbjEyMzQiXQ==',
          node: {
            followedBy: 2,
            id: 'kauppikseen1234',
            name: 'kauppikseen',
            questions: 4,
          },
        },
        {
          cursor: 'WyIyMDIyLTA0LTI4VDA3OjEyOjQ2LjI0MVoiLCJESUEyMDIyMTIzNCJd',
          node: {
            followedBy: 1,
            id: 'DIA20221234',
            name: 'DIA2022',
            questions: 1,
          },
        },
      ],
    },
  },
  {
    request: {
      query: GET_STACKS,
    },
    result: {
      data: [
        {
          cursor:
            'WyIyMDIyLTA1LTA3VDEyOjUwOjU5LjE3N1oiLCJrYXVwcGlzLXloMTIzNCJd',
          node: {
            id: 'kauppis-yh1234',
            name: 'kauppis-yh',
            public: false,
            questions: 3,
            followedBy: 0,
            createdById: 'bbe42984-051b-4a01-b45d-b8d29c32200c',
            createdAt: '2022-05-07T12:50:59.177Z',
            updatedAt: '2022-05-07T12:50:59.177Z',
            tags: [],
          },
        },
        {
          cursor: 'WyIyMDIyLTA1LTA3VDExOjUwOjU5LjE3N1oiLCJESUEyMDIyMTIzNCJd',
          node: {
            id: 'DIA20221234',
            name: 'DIA2022',
            public: false,
            questions: 0,
            followedBy: 0,
            createdById: 'bbe42984-051b-4a01-b45d-b8d29c32200c',
            createdAt: '2022-05-07T11:50:59.177Z',
            updatedAt: '2022-05-07T11:50:59.177Z',
            tags: [],
          },
        },
        {
          cursor:
            'WyIyMDIyLTA1LTA3VDEwOjUwOjU5LjE3N1oiLCJ5b21hdGVtYXRpaWtrYTEyMzQiXQ==',
          node: {
            id: 'yomatematiikka1234',
            name: 'yomatematiikka',
            public: false,
            questions: 0,
            followedBy: 0,
            createdById: 'bbe42984-051b-4a01-b45d-b8d29c32200c',
            createdAt: '2022-05-07T10:50:59.177Z',
            updatedAt: '2022-05-07T10:50:59.177Z',
            tags: [],
          },
        },
      ],
    },
  },
];

describe('rendering tests', () => {
  let component: RenderAPI;

  beforeEach(() => {
    component = render(
      <MockedProvider mocks={mocks} addTypename={true}>
        <Library />
      </MockedProvider>,
    );
  });

  test('should show loading indicator when data is initially loaded', async () => {
    await waitFor(() => {
      expect(component.getByText('Loading')).toBeTruthy;
    })
  });
  
  test('should render the flatlist when not loading', async () => {
    await waitFor(async () => {
      expect(component.getAllByTestId('library-list').length).toBe(1);
      expect(component).toMatchSnapshot();
    });
  })
});
