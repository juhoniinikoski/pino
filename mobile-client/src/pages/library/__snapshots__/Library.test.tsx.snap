// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering tests should render the flatlist when not loading 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "left",
      "right",
    ]
  }
  style={
    Array [
      Object {
        "display": "flex",
        "flex": 1,
        "marginHorizontal": 16,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="library-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RNCSafeAreaView>
`;
